src/Tools/BuildValidator/BuildConstructor.cs

```diff
@@ -59,53 +62,115 @@ public Compilation CreateCompilation(MetadataReader metadataReader, string name)
             throw new InvalidDataException("Did not find language in compilation options");
         }
 
-        private ImmutableArray<MetadataReference> CreateMetadataReferences(CompilationOptionsReader pdbReader)
+        private ImmutableArray<MetadataReferenceInfo> GetMetadataReferenceInfos(CompilationOptionsReader compilationOptionsReader)
+        {
+            var referenceInfos = compilationOptionsReader.GetMetadataReferences();
+            return referenceInfos;
+        }
+
+        private ImmutableArray<SourceFileInfo> GetSourceFileInfos(CompilationOptionsReader compilationOptionsReader, Encoding encoding)
+        {
+            return compilationOptionsReader.GetSourceFileInfos(encoding);
+        }
+
+        private ImmutableArray<MetadataReference> ResolveMetadataReferences(ImmutableArray<MetadataReferenceInfo> referenceInfos)
         {
-            var referenceInfos = pdbReader.GetMetadataReferences();
+            _logger.LogInformation("Locating metadata references");
             return _referenceResolver.ResolveReferences(referenceInfos);
         }
 
-        private ImmutableArray<SourceText> GetSources(CompilationOptionsReader pdbReader, Encoding encoding)
+        private ImmutableArray<SourceLink> ResolveSourceLinks(CompilationOptionsReader compilationOptionsReader)
         {
-            var builder = ImmutableArray.CreateBuilder<SourceText>();
+            using var _ = _logger.BeginScope("Source Links");
+            var sourceLinks = compilationOptionsReader.GetSourceLinksOpt();
+            if (sourceLinks.IsDefault)
+            {
+                _logger.LogWarning("No source links found in pdb");
```

Comment:
I would use `LogInformation` here because this is not necessarily a warning. Consider that its' completely valid for the entire source to just be embedded, no source link at all. That's actually the simplest possible configuration.

Reply:
Fixed